test/lib/Benchmark.js

```diff
@@ -0,0 +1,182 @@
+/* globals XPCOMUtils, Services, Task */
+"use strict";
+
+const {Cu} = require("chrome");
+Cu.import("resource://gre/modules/XPCOMUtils.jsm");
+Cu.import("resource://gre/modules/Services.jsm");
+
+XPCOMUtils.defineLazyModuleGetter(this, "Task",
+                                  "resource://gre/modules/Task.jsm");
+// For the performance.now()
+const performance = Services.appShell.hiddenDOMWindow.performance;
+
+/*
+ * A simple benchmark framework that runs on top of jpm test
+ *
+ * Note that the timing resolution for the benchmark is microsecond
+ *
+ * Example:
+
+  function fib(n) {
+    if (n < 2) {
+      return n;
+    }
+
+    return fib(n - 1) + fib(n - 2);
+  }
+
+  exports.test_benchmark = function*() {
+    let bench = new Benchmark(b => {
+      for (let i = 0; i < b.N; i++) {
+        fib(20);
+      }
+    });
+    let result = yield bench.launch();
+    console.log(result);
+  };
+*/
+
+/*
+ * {Params} func is the benchmark function/generator that is of the following form
+ *
+   function(b) {
+     for (let i = 0; i < b.N; i++) {
+       // do your stuff here
+     }
+   }
+ * Note that the framework will decide the best iteration number b.N based on
+ * how fast the target benchmark function runs
+ *
+ * {Params} benchTime is the benchmark time in microseconds, by default it's one second
+ */
+function Benchmark(func, benchTime = 1e6) {
+  this._failed = false;
+  this._benchFunc = func;
+  this._benchTime = benchTime;
+  this._timerOn = false;
+  this._start = 0;
+  this._duration = 0;
+  // N is public to the benchmark functions
+  this.N = 0;
+}
+
+Benchmark.prototype = {
+  /*
+   * Starts the timer for the benchmark. This function is called automatically
+   * before each benchmark. It also can be used to resuming timing after the
+   * stopTimer function call
+   */
+  startTimer() {
+    if (!this._timerOn) {
+      this._start = performance.now() * 1e3;
+      this._timerOn = true;
+    }
+  },
+
+  /*
+   * Resets the start timer if the timer is in use. It also zeros the benchmark's
+   * duration
+   */
+  resetTimer() {
+    if (this._timerOn) {
+      this._start = performance.now() * 1e3;
+    }
+    this._duration = 0;
+  },
+
+  /*
+   * Stops the timer if the timer is in use. It also calculates the benchmark's
+   * duration
+   */
+  stopTimer() {
+    if (this._timerOn) {
+      this._duration += (performance.now() * 1e3 - this._start);
+      this._timerOn = false;
+    }
+  },
+
+  _usPerOp() {
+    if (this.N <= 0) {
+      return 0;
+    }
+    return Math.round(this._duration / this.N);
+  },
+
+  /*
+   * Rounds the given integer up to a number of the form [1eX, 2eX, 3eX, 5eX, 1e(X+1)]
+   */
```

Comment:
We wanted to roundup the iteration number to [1eX, 2eX, 3eX, 5eX, 10eX], instead of letting it completely go wild. This somewhat mimics the traditional benchmark approach, where the benchmark designer specifies some frequently used iteration numbers beforehand.